### PR TITLE
[infer] Add Bloom inference policy and replaced methods

### PR DESCRIPTION
**Pull request by yuanheng-zhao**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 📝 What does this PR do?

- Add Bloom inference policy
- Add Bloom inference forwards replaced methods
- Support Bloom inference flow
- Fix file paths

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by yuanheng-zhao** [comment]
will add an inference engine and revise on bloom modelling.

File: colossalai/shardformer/inference/engine.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+from colossalai.cluster import ProcessGroupMesh
+from colossalai.shardformer import ShardConfig, ShardFormer
+from colossalai.shardformer.inference import BatchInferState, MemoryManager
+# from colossalai.shardformer.policies.bloom import BloomModelInferPolicy
+from colossalai.shardformer.policies.auto_policy import get_autopolicy
+
+DP_AXIS, PP_AXIS, TP_AXIS = 0, 1, 2
+
+
+class InferenceEngine:
```
Comment (tiandiao123) [inline colossalai/shardformer/inference/engine.py:21]: change name into TpInferEngine :

**Comment by tiandiao123** [comment]
how about move engine outsude shardformer? Like this? https://github.com/hpcaitech/ColossalAI/pull/4492/files

File: colossalai/shardformer/inference/engine.py (hunk 1 above)
Comment (yuanheng-zhao) [reply colossalai/shardformer/inference/engine.py:21]: Gotcha. Will do while revising a few bugs of bloom modelling.

Comment (yuanheng-zhao) [reply colossalai/shardformer/inference/engine.py]: Updated

**Review by tiandiao123 — APPROVED** [review]

File: colossalai/inference/tensor_parallel/modeling/bloom.py [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+            closest_power_of_2 = 2**math.floor(math.log2(n))
+            return (get_slopes_power_of_2(closest_power_of_2) +
+                    get_slopes(2 * closest_power_of_2)[0::2][:n - closest_power_of_2])
+
+    slopes = torch.Tensor(get_slopes(n_head))
+    head_alibi = slopes.to(dtype)
+    return head_alibi    # 1 * num_heads
+
+
+def generate_alibi_2(n_head, dtype=torch.float16):
```
Comment (Xu-Kai) [inline colossalai/inference/tensor_parallel/modeling/bloom.py:66]: What's difference between generate_alibi and generate_alibi_2?

Comment (yuanheng-zhao) [reply colossalai/inference/tensor_parallel/modeling/bloom.py:66]: revised

**Pull request merged**